### DISCUSSION TITLE
Berry Json serialisation method

### DISCUSSION THREAD

**Discussion by usrrsr** [arendst/Tasmota#21985] [Q&A]
Dear all,

Can any one suggest to form following json string (ie Serialize) in berry 

```
{
      "message": {
        "topic": "246F28B49F88",
        "notification": {
          "title": "This is Title",
          "body": "This is Body"
        },
        "data": {
          "story_id": "story_12345",
          "Hostname": "246F28B48072",
          "IPAddress": "192.168.0.101",
          "IP6Global": "tt",
          "IP6Local": "fe80::266f:28ff:feb4:9f88"
        }
      }
    }
```

I tried for 2 days with json module but no success

Thanks,

**Comment by sfromis** [comment]
How to form it depends on what you have already. If you go the other way, starting from the plain string you posted, you could try a `json.load(myjson)` which will result in a map structure with a couple of levels.

If you then take that map structure (or a similar one), and do a `json.dump(mymap)`, you'll get the valid JSON string representing the same data, but the fields will likely not be in the same order.

As testing, you could even do this to form such a map structure. Yes, it looks like JSON, but it is not a string, it is how you can create a map structure directly in the Berry language.
```
var mymap=
{
      "message": {
        "topic": "246F28B49F88",
        "notification": {
          "title": "This is Title",
          "body": "This is Body"
        },
        "data": {
          "story_id": "story_12345",
          "Hostname": "246F28B48072",
          "IPAddress": "192.168.0.101",
          "IP6Global": "tt",
          "IP6Local": "fe80::266f:28ff:feb4:9f88"
        }
      }
}
```

If you do a `print` of the map structure, you can see that what you get looks very similar to JSON, but with the `"` string delimiters instead being `'`. There are a few other differences between a map converted to a string, and JSON, therefore you should always use the `json.dump(mymap)` function to create JSON, unless you prefer to do the string-building yourself.

To programmatically build such a map structure, you can use the standard functions for handling maps with keys, noting that values in a map element can themselves be maps.

**Reply by usrrsr** [reply]
Thanks for replying 

No it's not simple text with hard coded values

Structure will be same but values.i want to update with values of variable declared in berry script, so it's to be form programically 

I can't get since I lack experience in berry

Can I replaced text values with variable as place holder

It will be appreciable to get snippet to accomplish this so that I can try it out


Regards

**Reply by sfromis** [reply]
Yes, I knew it was not to be hard-coded values, that's why I only mentioned this as "testing". Mainly for you to see how it works.

But if you build the right map structure, analogous to what you get when using the testing suggestion, you can get JSON again from that. And the map handling in berry allows you to build such a map.

If you take the testing example, and use a variable instead of each constant, you'll get a map structure with what you want, and then you can use `json.dump` to get to JSON. This is because that was **not** text, but Berry code (very much resembling JSON) constructing such.

**Comment by wjohn007** [comment]
[A01DynClass.zip](https://github.com/user-attachments/files/16665503/A01DynClass.zip)
Maybe the solution with dynamic classes will help. 
An example would be

```berry

# build the complex structure
content = DynClass()
content.topic = "246F28B49F88"
content.notification=DynClass()
content.notification.title="this is title"
content.notification.body="this is body"
content.data=DynClass()
content.data.story_id = "story_12345"
content.data.Hostname="246F28B48072"
content.data.IPAddress="192.168.0.101"
content.data.IP6Global="tt"
content.data.IP6Local="fe80::266f:28ff:feb4:9f88"

mymap = DynClass()
mymap.message=content

```

Following command creates a json-string
```berry
# convert to json
mymap.toJson()
```

with this result
```json
{
	"message": {
		"topic": "246F28B49F88",
		"data": {
			"IP6Global": "tt",
			"story_id": "story_12345",
			"IP6Local": "fe80::266f:28ff:feb4:9f88",
			"Hostname": "246F28B48072",
			"IPAddress": "192.168.0.101"
		},
		"notification": {
			"title": "this is title",
			"body": "this is body"
		}
	}
}
```

**Reply by sfromis** [reply]
For a simple case of wanting to create JSON, just writing the map directly in Berry is more straightforward than creating that class hierarchy. In both cases converting to JSON after creation.

To illustrate how Berry works, the line-by-line method just gets a bit longer than the dynamic map, like this:
```
content = {}
content['topic'] = "246F28B49F88"
content['notification']={}
content['notification']['title']="this is title"
content['notification']['body']="this is body"
content['data']={}
content['data']['story_id'] = "story_12345"
content['data']['Hostname']="246F28B48072"
content['data']['IPAddress']="192.168.0.101"
content['data']['IP6Global']="tt"
content['data']['IP6Local']="fe80::266f:28ff:feb4:9f88"
mymap = {'message':content}
```

Still, I'd recommend what I already posted about using the "testing" `mymap` assignment, and just using variables where relevant. You get very close to just writing the desired JSON directly as a Berry map, before converting to a string.

**Comment by usrrsr** [comment]
Thank you sfromis and wjohn007 for your help
Because of both of you I can now go ahead with stucked coding part
I will try both and revert back

Regards

**Reply by usrrsr** [reply]
> Thank you sfromis and wjohn007 for your help Because of both of you I can now go ahead with stucked coding part I will try both and revert back
> 
> Regards

Hi i tried both code, both methods worked superb, finally able to receive firebase notification on my app offline.
I also tried method suggested by sfromis  and it works below is code snippet for future referance if someone needs

 
```
  def SendNotificationTest(ID,Device,NoTFType,State,Title,SubText,TextList,Qos)  
    # SendNotificationTest(1,0,2,0,"Summary","Status : %s self.curDate","self.NotimsgList.tostring",1) 

    #var b = f'{{"message":{"topic":"246F28B49F88","data":{"IP6Global":"{ID:d}","story_id":"{NoTFType:i}","IP6Local":"fe80::266f:28ff:feb4:9f88","Hostname":"246F28B48072","IPAddress":"192.168.0.101"},"notification":{"body":"{Title}","title":"{SubText}"}}}}'
    #print(b) 
    #print(json.dump(b),'format')) 

    var mymap={
              "message": {
                "topic": "246F28B49F88",
                "notification": {
                  "title": Title,
                  "body": SubText
                },
                "data": {
                  "ID": int(ID),
                  "Device": int(Device),
                  "NoTFType": int(NoTFType),
                  "State": int(State),
                  "Title": Title,
                  "SubText": SubText               
                }
              }
        }
    print(json.dump(mymap)) 
    print(json.dump(mymap,'format')) 

  end

 SendNotificationTest(1,0,2,0,"Summary","Status : %s self.curDate","self.NotimsgList.tostring",1)
   
 #- Output
    {"message":{"topic":"246F28B49F88","data":{"State":0,"SubText":"Status : %s self.curDate","NoTFType":2,"Device":0,"ID":1,"Title":"Summary"},"notification":{"body":"Status : %s self.curDate","title":"Summary"}}}
    {
      "message": {
        "topic": "246F28B49F88",
        "data": {
          "State": 0,
          "SubText": "Status : %s self.curDate",
          "NoTFType": 2,
          "Device": 0,
          "ID": 1,
          "Title": "Summary"
        },
        "notification": {
          "body": "Status : %s self.curDate",
          "title": "Summary"
        }
      }
    }

-#
```


I also did little bit study i tried to implement it another way as below but it won't worked

```
var ID =1
var NoTFType=2
var Title ="Title"
var SubText ="SubText"
  var b = f'{{"message":{"topic":"246F28B49F88","data":{"IP6Global":"{ID:d}","story_id":"{NoTFType:i}","IP6Local":"fe80::266f:28ff:feb4:9f88","Hostname":"246F28B48072","IPAddress":"192.168.0.101"},"notification":{"body":"{Title}","title":"{SubText}"}}}}'
    print(b) 
    print(json.dump(b),'format')) 
```


Thanks for your help

Regards,

**Reply by sfromis** [reply]
The reason why the last non-working usage of a f-string failed is that you need to be very careful about using and doubling those `{}`. Each and every curly brace in the intended output needs to be doubled, and you did not do that.

Such is a reason why I did not even suggest that technique, when the map format gets you so close to just having the JSON directly as Berry code. Keeping track of both usages of `{}` at the same time has a high risk of tripping it up.
